docs/handbook/04_streams.md

```diff
@@ -127,13 +127,43 @@ Content-Type: text/vnd.turbo-stream.html; charset=utf-8
 
 This `messages/message` template partial can then also be used to re-render the message following an edit/update operation. Or to supply new messages created by other users over a WebSocket or a SSE connection. Being able to reuse the same templates across the whole spectrum of use is incredibly powerful, and key to reducing the amount of work it takes to create these modern, fast applications.
 
-
 ## Progressively Enhance When Necessary
 
 It's good practice to start your interaction design without Turbo Streams. Make the entire application work as it would if Turbo Streams were not available, then layer them on as a level-up. This means you won't come to rely on the updates for flows that need to work in native applications or elsewhere without them.
 
 The same is especially true for WebSocket updates. On poor connections, or if there are server issues, your WebSocket may well get disconnected. If the application is designed to work without it, it'll be more resilient.
 
+If a Turbo Drive visit needs to integrate with Turbo Streams, you can overwrite its [Accept][] header within a `turbo:before-fetch-request` event listener:
+
+```js
+// turbo_streams_controller.js
+import { Controller } from "stimulus"
+
+export default class extends Controller {
+  injectHeadersIntoNextRequest() {
+    const injectHeaders = (event) => {
+      const { headers } = event.detail.fetchOptions || {}
+
+      if (headers) {
+        headers.Accept = `text/vnd.turbo-stream.html, ${headers.Accept}`
+      }
+    }
+
+    addEventListener("turbo:before-fetch-request", injectHeaders, { once: true })
+  }
+}
+```
+
+Then, attach the controller to your page's `<body>` element and route `turbo:click` events on `<a>` elements to the `turbo-streams#injectHeadersIntoNextRequest` action:
+
+```html
+<body data-controller="turbo-streams">
+  <!-- ... -->
+  <a href="/streams-request" data-action="turbo:click->turbo-streams#injectHeadersIntoNextRequest">Load content via Turbo Streams</a>
```

Comment:
honestly, this solution looks a bit hacky to me 😅 wouldn't it be possible to have something like:
```html
<a href="/streams-request" data-turbo-stream="true">Load content via Turbo Streams</a>
```
?

Reply:
@santib I'm on the fence about built-in support through an attribute, since GET requests resulting in stream responses are a bit of an anti-pattern, and we've only recently removed support for them in [7.0.0.beta4](https://github.com/hotwired/turbo/releases/tag/v7.0.0-beta.4). 

Of course, I defer to @sstephenson and @javan to decide.

Reply:
Thanks for responding Sean. Why do you see it as anti-pattern? I see a very useful use case that is the ability to have dynamically repeatable fields within a form. In my head the "Add" button would be a link_to (GET request) to a turbo stream endpoint that returns the HTML of the field to be appended to the form. Thoughts?

Reply:
>  In my head the "Add" button would be a link_to (GET request) to a turbo stream endpoint that returns the HTML of the field to be appended to the form.

Could that be achieved by combining two forms: one to make the Turbo Stream requests, and one for the Form Submission?

```html
<form id="more-widget-fields" method="post" action="/widget-fields">
  <input type="hidden" name="widget[widget_specific_data]" value="...">
</form>

<form action="/widgets" method="post">
  <input type="text" name="widget[name][]">
  <input type="text" name="widget[name][]">

  <button type="submit" form="more-widget-fields" name="widget[form_specific_data]" value="...">More rows</button>

  <button type="submit">Submit</button>
</form>
```

By making one `<button>` reference the `<form id="more-widget-fields">` via the [`[form]` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-form), we can drive Turbo Stream updates through a `<button>` nested within the `<form>` that intends to create a resource (a "widget" in this example).

> Why do you see it as an anti-pattern? 

It would be possible for the behavior to be triggered by an `<a>` element, but that undercuts the standard HTML semantics that an `<a>` element references another resource and drives navigation, and that `<button>` elements drive behavior on the current page.

This is a pseudocode example, and might require some more thought. The `name`, `widget_specific_data`, and `form_specific_data` mentioned by the form are totally arbitrary and made up, but intend to demonstrate how submission data can come from the surrounding elements in the page, and could very well be generated by a form builder or Turbo Streams themselves.

Reply:
@seanpdoyle Turbo could also handle the dynamic form generation. It would be nice to have a built-in feature for "nested forms". By the way, I've implemented a Stimulus controller that does exactly that. https://gist.github.com/rainerborene/644da1be59ecb4f6c4c9637abf2622e0

Reply:
Thanks Sean, I get your point on `<a>` behavior, and I agree. But what if it's a `button_to .. method: :get` instead of an `<a>`? I think that would be the best solution for the nested fields use case. IMHO a `POST` doesn't make sense to load another field since it's not submitting any data 😅 and I think a `GET` would be more appropriate (even if it's not an anchor tag)

Reply:
> honestly, this solution looks a bit hacky to me 😅 wouldn't it be possible to have something like:
> 
> ```
> <a href="/streams-request" data-turbo-stream="true">Load content via Turbo Streams</a>
> ```
> 
> ?

I already tried to implement this (before I found this suggestion), however this is still not a solution because Turbo visit still takes over and updates the page's url. So we'd need a way to stop that visit and just do regular AJAXy fetch.

Reply:
@seanpdoyle is it possible that https://github.com/hotwired/turbo/pull/146 wouldn't be possible (or at least the example in there) if we can't respond with turbo streams to GET requests, or am I missing something?

Reply:
Another example of what wouldn't be possible in Hotwire if we don't allow GET requests to respond with a turbo stream https://htmx.org/examples/infinite-scroll/

Other solution would be to get https://github.com/hotwired/turbo/pull/146 that actually I like it more than streams. If that's the plan I'm 👍 then

Reply:
One possible workaround might involve a `<turbo-frame>` for "operations" that you can drive with an anchor element:

```html
<turbo-frame id="operations"></turbo-frame>

<div id="feed">
  <article id="post_1">...</article>
  <!-- ... -->
  <a href="/operations/posts?page=2" data-turbo-frame="operations">Load more</a>
</div>
```

Then, make sure that your response renders an HTML content type containing a  `<turbo-frame id="operations">` element that renders the `<turbo-stream>` elements to update the page:

```html
<!-- app/views/operations/posts/index.html.erb -->
<turbo-frame id="operations">
  <turbo-stream target="feed" action="append">
    <template><!-- ... --></template>
  </turbo-stream>
</turbo-frame>
```

The frame will navigate, setting its `src="/operations/posts?page=2"` and the `<turbo-stream>` elements in the response will operate and then remove themselves.

You could re-purpose a single frame for most operations on the page.

> Other solution would be to get [hotwired/turbo#146](https://github.com/hotwired/turbo/pull/146) merged. Actually, I like it more than streams. 

I agree! I think that is going to take a little while to hash out the rest of the design details. 

In the interim, I think [#59](https://github.com/hotwired/turbo/pull/59) is likely to ship first, which adds a `turbo:before-frame-render` event as a hook to modify a frame's contents so that you could roll your own rendering logic for appending and whatnot.

Reply:
I was able to get this to work using this suggestion with modals after `get` support didn't exist anymore the way I was using it.

https://gitlab.com/doug.stull/turbo_modal/-/compare/turbo-rails-0.5.8...turbo-rails-0.5.9